docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -63,248 +155,342 @@ audience. For instance:
+
+MCP servers **MUST NOT** complete public client flows on behalf of the MCP client (e.g., take 
+authorization codes and exchange them for tokens with a public client ID, and then pass the tokens 
+back to the client).
+
+Public client flows **MUST** be completed by MCP Client using OAuth 2.1 Authorization
+Code Flow with PKCE.
+
+#### 2.4.2 Flow Description
+
+The third-party authorization flow comprises these steps:
+
+1. MCP client requests data from the MCP server.
+1. MCP server responds with `HTTP 401 Unauthorized` and `WWW-Authenticate` header containing the
+   URL to the PRM document.
+1. MCP client retrieves and validates the PRM document.
+1. MCP client retrieves the AS endpoint from the PRM document.
+1. MCP client redirects user to the AS authorization endpoint. MCP client can include scopes that
+   are defined in the PRM document or other scopes, as supported by the MCP server.
+1. User authorizes with authorization server.
+1. Authorization server redirects back to MCP client with authorization code.
+1. MCP client exchanges the code for the requested tokens.
+1. MCP client stores the token.
+1. MCP client can now request data from the MCP server with the token attached in the `Authorization`
+   header.
 
 ```mermaid
 sequenceDiagram
-    participant C as Client
-    participant S as Server
-
-    C->>S: GET /.well-known/oauth-authorization-server
-    alt Discovery Success
-        S->>C: 200 OK + Metadata Document
-        Note over C: Use endpoints from metadata
-    else Discovery Failed
-        S->>C: 404 Not Found
-        Note over C: Fall back to default endpoints
-    end
-    Note over C: Continue with authorization flow
+    participant UA as User-Agent (Browser)
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+    Note over MC: MCP Client retrieves and validates<br/>PRM document
+    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
+    MC-->>UA: Initiate authorization request at the authorization endpoint
+    Note over UA: User authenticates and authorizes<br/>client for scope and resource
+    UA->>MC: Return authorization code to client
+    MC->>AS: Exchange code for token
+    AS-->>MC: Access token
+    MC->>MC: Cache token
+    MC->>MS: Request data</br>(with Authorization header)
 ```
 
-#### 2.3.1 Server Metadata Discovery Headers
+#### 2.4.4 Security Considerations
+
+When implementing authorization with an existing AS, MCP servers **MUST**:
 
-MCP clients _SHOULD_ include the header `MCP-Protocol-Version: <protocol-version>` during
-Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol
-version.
+1. Request the minimum number of required scopes.  
+2. Securely store any tokens obtained from the client and/or the authorization server.  
```

Comment:
Sorry to kick this one open again - the phrasing here makes me think that the MCP server is required to store the credentials from the client, but that shouldn't be a requirement.

Would this line be better under `2.5.1`? In the standard delegated AS model the MCP server can be made stateless - acquire token from request; validate token; return response. Only with delegated OAuth does the MCPS retrieve a token directly from the authorization server.

The token exchange example provided previously is a server-to-server authorization pattern and outside the scope of this spec.

Reply:
I reframed this a bit to be more generic and less prescriptive.